### PR TITLE
Adding conversion between LinearLayout and DPASLayout

### LINKED ISSUE #1295: [test_convert2d] Fix test failures on XPU
Upstream commit https://github.com/triton-lang/triton/commit/1e797c4f3f73fa2c90933afeda5f554d7f6d72ee (merged in https://github.com/intel/intel-xpu-backend-for-triton/pull/1294) enabled some `test_convert2d` tests, but they are still failing on XPU.

To reproduce, remove from `scripts/skiplist/default/language.txt`.

### PR DESCRIPTION
**Pull request by hwnam831**
This PR fixes #1295.

Upstream Triton tries to abstract all tensor layouts with the LinearLayout abstraction.
Solving #1295 requires using LinearLayout, so I've added codes to convert DPASLayout to LinearLayout.
Currently, the conversion is only correct for DPAS OperandC layout.

### CODE REVIEW

File: third_party/intel/include/Dialect/TritonIntelGPU/IR/LinearLayoutConversions.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+namespace mlir::triton::gpu {
+
+// - DPASLayout has three derivatives
+//
+//   OperandA (opidx==0)
+//   OperandB (opidx==1)
+//   OperandC (no opidx. default to -1)
+//
+std::optional<LinearLayout> DPAStoLinearLayout(ArrayRef<int64_t> shape,
+                                               Attribute layout, int opidx = 2);
```
Comment (whitneywhtsang) [inline third_party/intel/include/Dialect/TritonIntelGPU/IR/LinearLayoutConversions.h:21]: In the comment, it is written that default to -1, but here it is default to 2.

File: third_party/intel/lib/Dialect/TritonIntelGPU/IR/CMakeLists.txt [hunk 2]
```diff
@@ -1,11 +1,12 @@
...
   Ops.cpp
-
+  LinearLayoutConversions.cpp
   DEPENDS
   TritonIntelGPUTableGen
   TritonIntelGPUAttrDefsIncGen
 
   LINK_LIBS PUBLIC
   MLIRSideEffectInterfaces
+
```
Comment (whitneywhtsang) [inline third_party/intel/lib/Dialect/TritonIntelGPU/IR/CMakeLists.txt:11]: Remove.

File: third_party/intel/lib/TritonIntelGPUToLLVM/Utility.h [hunk 3]
```diff
@@ -747,19 +754,32 @@ loadSharedToDistributed(Value dst, Value src, SharedMemoryObject &shrMemObj,
...
+                 target, [&](VectorType vecTy, Value vecAddr) {
+                   auto vecVal = load(vecTy, vecAddr);
+                   vecVal.setAlignment(vecTy.getNumElements() *
+                                       elemTy.getIntOrFloatBitWidth() / 8);
+                   for (int v = 0; v < vecTy.getNumElements(); v++) {
+                     ret.push_back(extract_element(elemTy, vecVal, i32_val(v)));
+                   }
+                 })) {
     return ret;
+  }
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/Utility.h:782]:
```suggestion
  }
  bool success = emitTransferBetweenRegistersAndShared(
                 dstTy, srcTy, elemTy, /*maxVecElems=*/std::nullopt,
                 shrMemObj.getBase(), shrMemObj.getStrides(), loc, rewriter,
                 target, [&](VectorType vecTy, Value vecAddr) {
                   auto vecVal = load(vecTy, vecAddr);
                   vecVal.setAlignment(vecTy.getNumElements() *
                                       elemTy.getIntOrFloatBitWidth() / 8);
                   for (int v = 0; v < vecTy.getNumElements(); v++) {
                     ret.push_back(extract_element(elemTy, vecVal, i32_val(v)));
                   }
                 });
  if (!success)
    llvm::report_fatal_error("Failed to emit transfer from shared to register");
```
and then remove the code after.

**Review by whitneywhtsang — COMMENTED** [review]
initial review comments.

File: third_party/intel/include/Dialect/TritonIntelGPU/IR/LinearLayoutConversions.h (hunk 1 above)
Comment (hwnam831) [reply third_party/intel/include/Dialect/TritonIntelGPU/IR/LinearLayoutConversions.h:21]: Fixed

File: third_party/intel/lib/TritonIntelGPUToLLVM/Utility.cpp [hunk 4]
```diff
@@ -140,3 +142,142 @@ Value llPrintf(ConversionPatternRewriter &rewriter, StringRef msg,
...
+  StringAttr kBlock = str_attr("block");
+  StringAttr kRegister = str_attr("register");
+  StringAttr kLane = str_attr("lane");
+  StringAttr kWarp = str_attr("warp");
+
+  std::optional<LinearLayout> regLayout;
+  if (auto dpasLayout = dyn_cast<DpasEncodingAttr>(registerTy.getEncoding())) {
+    // Check whether the layout specifies the opidx.
+    unsigned opIdx = 2;
+    if (triton::gpu::intel::hasDotDpasEncoding(registerTy)) {
```
Comment (chengjunlu) [inline third_party/intel/lib/TritonIntelGPUToLLVM/Utility.cpp:167]: This branch is taken only when the layout is the DPAS layout. 
It seems this condition never hit.

File: third_party/intel/lib/TritonIntelGPUToLLVM/Utility.cpp [hunk 5]
```diff
@@ -140,3 +142,142 @@ Value llPrintf(ConversionPatternRewriter &rewriter, StringRef msg,
...
+          triton::gpu::intel::getDotEncoding(registerTy).value();
+      opIdx = dotLayout.getOpIdx();
+    }
+    regLayout = triton::gpu::DPAStoLinearLayout(shape, dpasLayout, opIdx);
+  } else {
+    regLayout = triton::gpu::toLinearLayout(shape, registerTy.getEncoding());
+  }
+
+  std::optional<LinearLayout> sharedLayout;
+  if (auto dpasLayout = dyn_cast<DpasEncodingAttr>(sharedTy.getEncoding())) {
```
Comment (chengjunlu) [inline third_party/intel/lib/TritonIntelGPUToLLVM/Utility.cpp:178]: The tensor in SLM should always have the SharedLayout. 
It seems this condition is always false.

File: third_party/intel/lib/TritonIntelGPUToLLVM/Utility.cpp [hunk 6]
```diff
@@ -140,3 +142,142 @@ Value llPrintf(ConversionPatternRewriter &rewriter, StringRef msg,
 }
 
 } // namespace mlir::LLVM::intel
+
+namespace mlir::triton::intel {
+bool emitTransferBetweenDPASAndShared(
```
Comment (chengjunlu) [inline third_party/intel/lib/TritonIntelGPUToLLVM/Utility.cpp:147]: The LinearLayout allow us to use the common lowering code in TritonGPUToLLVM.

Is there any limitation in the function for us to re-use.
https://github.com/triton-lang/triton/blob/9c7f9f8e3f6c769d0c32d0c51265c04ba8b164ca/lib/Conversion/TritonGPUToLLVM/Utility.cpp#L267C6-L267C43

**Comment by chengjunlu** [comment]
The LinearLayout conversion is complicated.
Please add the unit test like the upstream test cases for NV MMA.
https://github.com/triton-lang/triton/blob/9c7f9f8e3f6c769d0c32d0c51265c04ba8b164ca/unittest/Dialect/TritonGPU/LinearLayoutConversionsTest.cpp#L291

File: third_party/intel/lib/Dialect/TritonIntelGPU/IR/LinearLayoutConversions.cpp [hunk 7]
```diff
@@ -0,0 +1,540 @@
...
+//
+// IMO the name CTALayoutAttr is wrong.  If we tried to be consistent anyway,
+// then we'd have to rename ctaLayout to "warpLayout".  I think that's more
+// confusing than being inconsistent about "cgaLayout", especially when we have
+// to consider the size of the warpLayout (surely that's not the "warpSize").
+
+#define S(v) StringAttr::get(ctx, (v))
+
+// Returns ["out0", "out1", ..., "out<rank-1>"].
+SmallVector<StringAttr> standardOutDimNames(MLIRContext *ctx, int rank) {
```
Comment (chengjunlu) [inline third_party/intel/lib/Dialect/TritonIntelGPU/IR/LinearLayoutConversions.cpp:37]: Let's ask Triton whether those common utils can be exported for the third party code to re-use.

File: third_party/intel/lib/TritonIntelGPUToLLVM/Utility.cpp (hunk 4 above)
Comment (hwnam831) [reply third_party/intel/lib/TritonIntelGPUToLLVM/Utility.cpp]: It is fixed. Now it separates the case of hasDotDpasEncoding and isa<DpasEncodingAttr> in emitTransferBetweenDPASandShared.

File: third_party/intel/lib/TritonIntelGPUToLLVM/Utility.cpp (hunk 5 above)
Comment (hwnam831) [reply third_party/intel/lib/TritonIntelGPUToLLVM/Utility.cpp]: emitTransferBetweenRegisterAndShared is bi-directional method. It also handles the case when registerTy is shared and sharedTy is RankedTensorType. This is how the upstream handles this function.

File: third_party/intel/lib/TritonIntelGPUToLLVM/Utility.cpp (hunk 6 above)
Comment (hwnam831) [reply third_party/intel/lib/TritonIntelGPUToLLVM/Utility.cpp:147]: I have been avoiding modifying the upstream part of the codes.
If we modify toLinearLayout (lib/Dialect/TritonGPU/IR/LinearLayoutConversion.cpp) directly, everything would become a lot simpler.

**Comment by hwnam831** [comment]
> The LinearLayout conversion is complicated. Please add the unit test like the upstream test cases for NV MMA. https://github.com/triton-lang/triton/blob/9c7f9f8e3f6c769d0c32d0c51265c04ba8b164ca/unittest/Dialect/TritonGPU/LinearLayoutConversionsTest.cpp#L291

I have added the unit tests under `unittest/Dialect/TritonGPU/DPAStoLinearLayoutTest.cpp`

**Review by whitneywhtsang — APPROVED** [review]

**Review by chengjunlu — APPROVED** [review]

**Pull request merged**